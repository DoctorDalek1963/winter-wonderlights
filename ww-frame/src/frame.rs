//! This module handles the implementations of the frames.

use crate::{FrameObject, RGBArray};
use serde::{Deserialize, Serialize};
use std::fmt;
use tracing::trace;
use ww_gift_coords::COORDS;

/// A type of frame data.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum FrameType {
    /// A frame indicating that all the lights are off.
    Off,

    /// The colour for each light on the chain.
    ///
    /// The driver is expected to handle the cases where the Vec is too long or too short and
    /// doesn't corespond one-to-one to the lights, but this behaviour is specific to individual
    /// drivers and should not be relied on.
    RawData(Vec<RGBArray>),

    /// A sophisticated 3D frame, made of several objects.
    Frame3D(Frame3D),
}

/// A 3D frame, made of several objects.
#[derive(Clone, PartialEq, Serialize, Deserialize)]
pub struct Frame3D {
    /// The vec of objects in the frame.
    ///
    /// An object later in the vec will override any object previous in the list. No blending of
    /// colours is performed.
    objects: Vec<FrameObject>,

    /// Whether to blend objects together.
    blend: bool,

    /// Optionally pre-computed data. If an effect needs to compute the raw data first, then it can
    /// just store that data here so the driver doesn't have to re-compute it later.
    #[cfg_attr(any(feature = "insta", test), serde(skip))]
    pre_computed_raw_data: Option<Vec<RGBArray>>,
}

impl fmt::Debug for Frame3D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Frame3D")
            .field("objects", &self.objects)
            .field("blend", &self.blend)
            //.field("pre_computed_raw_data", &self.pre_computed_raw_data)
            .finish()
    }
}

impl Frame3D {
    /// Create a new frame.
    pub fn new(objects: Vec<FrameObject>, blend: bool) -> Self {
        Self {
            objects,
            blend,
            pre_computed_raw_data: None,
        }
    }

    /// Compute the vec of raw data for this frame, using [`struct@COORDS`] to know where the
    /// lights are. Use [`Self::to_raw_data`] to get the data out or [`Self::raw_data`] to
    /// reference the optional data.
    ///
    /// This function does nothing if the data was already computed.
    ///
    /// This function returns a reference to self to allow things like:
    /// ```
    /// #![feature(let_chains)]
    ///
    /// # use ww_frame::Frame3D;
    /// # fn do_something(_: &Vec<[u8; 3]>) -> bool { true }
    /// # let mut frame = Frame3D::new(vec![], false);
    /// while let Some(data) = frame.compute_raw_data().raw_data() && do_something(data) {
    ///     // Do stuff
    ///     # break;
    /// }
    /// ```
    /// Admittedly this API isn't very ergonomic and this function should ideally return a
    /// reference to the computed data, but the borrow checker didn't like that, so we're currently
    /// using this bodge.
    pub fn compute_raw_data(&mut self) -> &mut Self {
        if self.pre_computed_raw_data.is_some() {
            return self;
        }

        let mut data: Vec<RGBArray> = vec![[0, 0, 0]; COORDS.lights_num()];
        trace!(?data, "Before");

        if self.blend {
            let total = self.objects.len() as f32;

            data = self
                .objects
                .iter()
                // Render the data into its own slice and convert all the u8s to f32s for later
                .map(|f_obj| {
                    let mut new_data = data.clone();
                    f_obj.render_into_slice(&mut new_data);
                    new_data
                        .into_iter()
                        // Divide by the total here to make the mean easier to calculate
                        .map(|[r, g, b]| [r as f32 / total, g as f32 / total, b as f32 / total])
                        .collect::<Vec<_>>()
                })
                // Sum all the data to get one Vec<[f32; 3]> where each element is the mean colour
                // for that light
                .fold(vec![[0., 0., 0.]; COORDS.lights_num()], |mut acc, v| {
                    for i in 0..COORDS.lights_num() {
                        acc[i][0] += v[i][0];
                        acc[i][1] += v[i][1];
                        acc[i][2] += v[i][2];
                    }
                    acc
                })
                .into_iter()
                // Convert everything back to u8s
                .map(|[r, g, b]| [r as u8, g as u8, b as u8])
                .collect();
        } else {
            for frame_object in &self.objects {
                frame_object.render_into_slice(&mut data);
            }
        }

        trace!(?data, "After");
        self.pre_computed_raw_data = Some(data);
        self
    }

    /// Return the raw data for this frame.
    pub fn to_raw_data(mut self) -> Vec<RGBArray> {
        self.compute_raw_data();
        self.pre_computed_raw_data.expect(
            "pre_computed_raw_data must be populated since compute_raw_data() has been called",
        )
    }

    /// Get an optional reference to the internal pre-computed data.
    pub fn raw_data(&self) -> Option<&Vec<RGBArray>> {
        self.pre_computed_raw_data.as_ref()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Object;
    use glam::Vec3;

    #[test]
    fn to_raw_data_test() {
        let green_plane = FrameObject {
            object: Object::Plane {
                normal: Vec3::new(1., 0.5, -3.5).normalize(),
                k: -1.2354,
                threshold: 0.15,
            },
            colour: [25, 200, 16],
            fadeoff: 0.14,
        };
        let blue_plane = FrameObject {
            object: Object::Plane {
                normal: Vec3::new(-1.6, 1.99, -0.02).normalize(),
                k: 0.24,
                threshold: 0.3,
            },
            colour: [0, 12, 186],
            fadeoff: 0.82,
        };
        let red_sphere = FrameObject {
            object: Object::Sphere {
                center: Vec3::new(0.01, 0.31, 0.2),
                radius: 1.2,
            },
            colour: [243, 36, 17],
            fadeoff: 0.1,
        };

        // Single green plane

        let single_green_plane = Frame3D::new(vec![green_plane.clone()], false);
        insta::with_settings!({
            info => &single_green_plane,
            description => "Rendering a single plane with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(single_green_plane.to_raw_data());
        });

        // Green and blue planes

        let green_and_blue_planes_no_blend =
            Frame3D::new(vec![green_plane.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &green_and_blue_planes_no_blend,
            description => "Rendering two planes with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(green_and_blue_planes_no_blend.to_raw_data());
        });

        let green_and_blue_planes_with_blend =
            Frame3D::new(vec![green_plane.clone(), blue_plane.clone()], true);
        insta::with_settings!({
            info => &green_and_blue_planes_with_blend,
            description => "Rendering two planes with with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(green_and_blue_planes_with_blend.to_raw_data());
        });

        // Blue plane and red sphere

        let blue_plane_red_sphere_no_blend =
            Frame3D::new(vec![blue_plane.clone(), red_sphere.clone()], false);
        insta::with_settings!({
            info => &blue_plane_red_sphere_no_blend,
            description => "Rendering a blue plane and then a red sphere with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(blue_plane_red_sphere_no_blend.to_raw_data());
        });

        let red_sphere_blue_plane_no_blend =
            Frame3D::new(vec![red_sphere.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &red_sphere_blue_plane_no_blend,
            description => "Rendering a red sphere and then a blue plane with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(red_sphere_blue_plane_no_blend.to_raw_data());
        });

        let blue_plane_red_sphere_with_blend =
            Frame3D::new(vec![blue_plane.clone(), red_sphere.clone()], false);
        insta::with_settings!({
            info => &blue_plane_red_sphere_with_blend,
            description => "Rendering a blue plane and then a red sphere with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(blue_plane_red_sphere_with_blend.to_raw_data());
        });

        let red_sphere_blue_plane_with_blend =
            Frame3D::new(vec![red_sphere.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &red_sphere_blue_plane_with_blend,
            description => "Rendering a red sphere and then a blue plane with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(red_sphere_blue_plane_with_blend.to_raw_data());
        });
    }
}
