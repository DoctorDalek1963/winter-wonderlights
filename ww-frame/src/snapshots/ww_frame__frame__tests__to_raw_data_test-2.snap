---
source: ww-frame/src/frame.rs
description: Rendering two planes with no blend
info:
  objects:
    - object:
        Plane:
          normal:
            - 0.27216554
            - 0.13608277
            - -0.9525794
          k: -1.2354
          threshold: 0.15
      colour:
        - 25
        - 200
        - 16
      fadeoff: 0.14
    - object:
        Plane:
          normal:
            - -0.6265843
            - 0.7793142
            - -0.007832304
          k: 0.24
          threshold: 0.3
      colour:
        - 0
        - 12
        - 186
      fadeoff: 0.82
  blend: false
---
[
  (0, 12, 186),
  (0, 10, 165),
  (0, 11, 176),
  (0, 10, 158),
  (0, 11, 178),
  (0, 10, 165),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 10, 162),
  (0, 11, 182),
  (0, 10, 169),
  (0, 10, 165),
  (0, 10, 159),
  (0, 11, 176),
  (0, 11, 184),
  (0, 11, 185),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 184),
  (0, 11, 184),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 10, 155),
  (0, 10, 165),
  (0, 9, 148),
  (0, 9, 140),
  (0, 7, 121),
  (0, 6, 100),
  (0, 6, 105),
  (0, 7, 120),
  (0, 5, 83),
  (0, 1, 22),
  (0, 0, 11),
  (0, 2, 33),
  (0, 2, 45),
  (0, 4, 64),
  (0, 3, 58),
  (0, 6, 107),
  (0, 9, 153),
  (0, 9, 143),
  (0, 8, 130),
  (0, 11, 171),
  (0, 12, 186),
  (0, 10, 163),
  (0, 8, 135),
  (0, 8, 136),
  (0, 9, 140),
  (0, 6, 94),
  (0, 6, 100),
  (0, 4, 63),
  (0, 5, 85),
  (0, 6, 99),
  (0, 5, 92),
  (0, 2, 39),
  (0, 0, 12),
  (0, 3, 58),
  (0, 0, 1),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 1, 24),
  (0, 3, 61),
  (0, 5, 86),
  (0, 6, 98),
  (0, 7, 113),
  (0, 9, 141),
  (0, 9, 150),
  (0, 12, 186),
  (0, 9, 140),
  (0, 7, 121),
  (0, 7, 110),
  (0, 5, 79),
  (0, 4, 64),
  (0, 4, 63),
  (0, 0, 3),
  (0, 0, 0),
  (0, 3, 50),
  (0, 1, 30),
  (0, 3, 54),
  (0, 4, 72),
  (0, 4, 75),
  (0, 7, 118),
  (0, 7, 110),
  (0, 7, 122),
  (0, 10, 168),
  (0, 7, 121),
  (0, 8, 126),
  (0, 8, 138),
  (0, 6, 101),
  (0, 9, 150),
  (0, 10, 159),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 177),
  (0, 9, 149),
  (0, 12, 186),
  (0, 10, 166),
  (0, 9, 145),
  (0, 6, 102),
  (0, 7, 109),
  (0, 5, 81),
  (0, 4, 74),
  (0, 7, 108),
  (0, 4, 77),
  (0, 6, 103),
  (0, 9, 145),
  (0, 8, 131),
  (0, 10, 159),
  (0, 7, 113),
  (0, 8, 138),
  (0, 8, 134),
  (0, 6, 93),
  (0, 8, 130),
  (0, 9, 148),
  (0, 8, 134),
  (0, 10, 161),
  (0, 6, 105),
  (0, 5, 89),
  (0, 6, 100),
  (0, 9, 151),
  (0, 8, 138),
  (0, 11, 176),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 176),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 171),
  (0, 10, 159),
  (0, 7, 115),
  (0, 6, 96),
  (0, 6, 94),
  (0, 8, 127),
  (0, 5, 88),
  (0, 8, 133),
  (0, 6, 106),
  (0, 9, 141),
  (0, 7, 108),
  (0, 7, 111),
  (0, 7, 122),
  (0, 9, 141),
  (0, 8, 135),
  (0, 8, 134),
  (0, 6, 94),
  (0, 8, 137),
  (0, 10, 160),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 9, 139),
  (0, 9, 151),
  (0, 6, 103),
  (0, 7, 117),
  (0, 7, 113),
  (0, 9, 150),
  (0, 9, 153),
  (0, 10, 163),
  (0, 12, 186),
  (0, 11, 176),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 178),
  (0, 12, 186),
  (0, 12, 186),
  (0, 10, 157),
  (0, 12, 186),
  (0, 11, 172),
  (0, 12, 186),
  (0, 10, 170),
  (0, 10, 166),
  (0, 7, 122),
  (0, 8, 124),
  (0, 7, 114),
  (0, 7, 110),
  (0, 5, 91),
  (0, 7, 109),
  (0, 5, 91),
  (0, 4, 66),
  (0, 7, 109),
  (0, 12, 186),
  (0, 5, 92),
  (0, 5, 86),
  (0, 5, 92),
  (0, 5, 82),
  (0, 8, 139),
  (0, 6, 95),
  (0, 9, 150),
  (0, 7, 123),
  (0, 9, 144),
  (0, 6, 108),
  (0, 9, 140),
  (0, 8, 138),
  (0, 8, 133),
  (0, 8, 132),
  (0, 7, 117),
  (0, 8, 126),
  (0, 7, 117),
  (0, 7, 120),
  (0, 8, 127),
  (0, 7, 113),
  (0, 4, 76),
  (0, 3, 46),
  (0, 2, 35),
  (0, 3, 55),
  (0, 3, 56),
  (0, 1, 16),
  (0, 3, 50),
  (0, 2, 39),
  (0, 1, 28),
  (0, 1, 21),
  (0, 4, 63),
  (0, 1, 25),
  (0, 1, 23),
  (0, 0, 4),
  (0, 0, 0),
  (0, 0, 6),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 1, 23),
  (0, 0, 4),
  (0, 6, 93),
  (0, 3, 58),
  (0, 4, 63),
  (0, 5, 83),
  (0, 6, 93),
  (0, 5, 88),
  (0, 5, 87),
  (0, 5, 84),
  (0, 5, 81),
  (0, 5, 86),
  (0, 6, 101),
  (0, 2, 40),
  (0, 2, 31),
  (0, 4, 67),
  (0, 4, 69),
  (0, 7, 112),
  (0, 6, 103),
  (0, 5, 84),
  (0, 6, 100),
  (0, 4, 69),
  (0, 2, 33),
  (0, 3, 58),
  (0, 4, 63),
  (0, 4, 75),
  (0, 2, 43),
  (0, 4, 70),
  (0, 1, 22),
  (0, 2, 37),
  (0, 2, 42),
  (0, 0, 0),
  (0, 2, 43),
  (0, 3, 48),
  (0, 3, 60),
  (0, 3, 49),
  (0, 4, 69),
  (0, 5, 78),
  (0, 3, 58),
  (0, 7, 122),
  (0, 8, 126),
  (0, 8, 124),
  (0, 8, 133),
  (0, 7, 121),
  (0, 10, 160),
  (0, 7, 113),
  (0, 5, 77),
  (0, 8, 129),
  (0, 8, 128),
  (0, 5, 91),
  (0, 8, 134),
  (0, 9, 148),
  (0, 9, 146),
  (0, 7, 121),
  (0, 8, 130),
  (0, 8, 137),
  (0, 11, 177),
  (0, 10, 159),
  (0, 11, 180),
  (0, 12, 186),
  (0, 9, 150),
  (0, 12, 186),
  (0, 12, 186),
  (0, 9, 151),
  (0, 11, 172),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 182),
  (0, 7, 119),
  (0, 8, 131),
  (0, 9, 150),
  (0, 7, 113),
  (0, 7, 119),
  (0, 5, 92),
  (0, 7, 115),
  (0, 7, 118),
  (0, 8, 130),
  (0, 9, 141),
  (0, 10, 158),
  (0, 10, 160),
  (0, 7, 115),
  (0, 7, 118),
  (0, 7, 123),
  (0, 10, 162),
  (0, 11, 174),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 9, 154),
  (0, 9, 150),
  (0, 8, 135),
  (0, 11, 182),
  (0, 11, 181),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 10, 159),
  (0, 10, 155),
  (0, 8, 127),
  (0, 9, 141),
  (0, 9, 146),
  (0, 8, 137),
  (0, 8, 136),
  (0, 9, 151),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 9, 149),
  (0, 10, 169),
  (0, 8, 126),
  (0, 8, 128),
  (0, 6, 105),
  (0, 5, 86),
  (0, 6, 100),
  (0, 5, 88),
  (0, 4, 72),
  (0, 5, 85),
  (0, 4, 75),
  (0, 8, 126),
  (0, 6, 96),
  (0, 8, 126),
  (0, 10, 163),
  (0, 8, 129),
  (0, 9, 141),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 176),
  (0, 10, 158),
  (0, 10, 163),
  (0, 9, 152),
  (0, 8, 134),
  (0, 8, 133),
  (0, 8, 128),
  (0, 8, 135),
  (0, 6, 104),
  (0, 10, 159),
  (0, 11, 173),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 10, 169),
  (0, 9, 140),
  (0, 11, 181),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 12, 186),
  (0, 11, 172),
  (0, 12, 186),
  (0, 10, 166),
  (0, 8, 127),
  (0, 7, 113),
  (0, 5, 86),
  (0, 7, 121),
  (0, 5, 86),
  (0, 5, 90),
  (0, 4, 67),
  (0, 5, 92),
  (0, 5, 78),
]
